pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -296,15 +298,301 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
+//
+//  1. Compute sandbox and container changes.
+//  2. Kill pod sandbox if necessary.
+//  3. Kill any containers that should not be running.
+//  4. Create sandbox if necessary.
+//  5. Create necessary containers
+//
+// TODO: support init containers in SyncPod.
+func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
+	podContainerChanges := m.computePodContainerChanges(pod, podStatus)
+	glog.V(3).Infof("computePodContainerChanges got %+v for pod %q", podContainerChanges, format.Pod(pod))
+	if podContainerChanges.CreateSandbox {
+		ref, err := api.GetReference(pod)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q: '%v'", format.Pod(pod), err)
+		}
+		m.recorder.Eventf(ref, api.EventTypeNormal, "SandboxChanged", "Pod sandbox changed, it will be killed and re-created.")
+	}
+
+	// Kill the pod if pod spec has changed.
+	if podContainerChanges.CreateSandbox || (len(podContainerChanges.ContainersToKeep) == 0 && len(podContainerChanges.ContainersToStart) == 0) {
+		if len(podContainerChanges.ContainersToKeep) == 0 && len(podContainerChanges.ContainersToStart) == 0 {
+			glog.V(4).Infof("Killing PodSandbox for %q because all other containers are dead.", format.Pod(pod))
+		} else {
+			glog.V(4).Infof("Killing PodSandbox for %q, will start new one", format.Pod(pod))
+		}
+
+		killResult := m.killPodWithSyncResult(pod, kubecontainer.ConvertPodStatusToRunningPod(podStatus), nil)
+		result.AddPodSyncResult(killResult)
+		if killResult.Error() != nil {
+			glog.Errorf("killPodWithSyncResult failed: %v", killResult.Error())
+			return
+		}
+	} else {
+		// Otherwise kill any running containers in this pod which are not to keep
+		runningContainerStatues := podStatus.GetRunningContainerStatuses()
+		for _, containerStatus := range runningContainerStatues {
+			if _, keep := podContainerChanges.ContainersToKeep[containerStatus.ID]; !keep {
+				glog.V(3).Infof("Killing unwanted container %q(id=%q) for pod %q", containerStatus.Name, containerStatus.ID, format.Pod(pod))
+
+				var podContainer *api.Container
+				var killMessage string
+				for i, c := range pod.Spec.Containers {
+					if c.Name == containerStatus.Name {
+						podContainer = &pod.Spec.Containers[i]
+						killMessage = podContainerChanges.ContainersToStart[i]
+						break
+					}
+				}
+				killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, containerStatus.Name)
+				result.AddSyncResult(killContainerResult)
+				if err := m.killContainer(pod, containerStatus.ID, podContainer, killMessage, nil); err != nil {
+					killContainerResult.Fail(kubecontainer.ErrKillContainer, err.Error())
+					glog.Errorf("killContainer %q(id=%q) for pod %q failed: %v", containerStatus.Name, containerStatus.ID, format.Pod(pod), err)
+					return
+				}
+			}
+		}
+	}
```

Comment:
The whole if-else logic is over-complicated in my opinion. This is basically because the result of `computePodContainerChanges` doesn't provide straightforward information, thus it requires us to compute again.

So how about we modify the `computePodContainerChanges` so that it returns:

- Whether a pod sandbox needs to be killed (let's count the "containers to keep, containers to start as well)
- Whether any containers needs to be killed
- Whether a pod sandbox needs to be started  (let's count the "containers to keep, containers to start as well)
- Whether any containers needs to be started

And then SyncPod should be much simpler and cleaner.